src/main/java/com/nvidia/spark/rapids/jni/kudo/MultiKudoTableVisitor.java

```diff
@@ -0,0 +1,264 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids.jni.kudo;
+
+import ai.rapids.cudf.HostMemoryBuffer;
+import ai.rapids.cudf.Schema;
+import com.nvidia.spark.rapids.jni.schema.SchemaVisitor;
+
+import java.util.*;
+
+import static com.nvidia.spark.rapids.jni.Preconditions.ensure;
+import static com.nvidia.spark.rapids.jni.kudo.KudoSerializer.padForHostAlignment;
+import static java.lang.Math.toIntExact;
+
+/**
+ * This class provides a base class for visiting multiple kudo tables, e.g. it helps to maintain internal states during
+ * visiting multi kudo tables, which makes it easier to do some calculations based on them.
+ * <br/>
+ * The schema used when visiting these kudo tables must be same as the schema used when creating these kudo tables.
+ */
+abstract class MultiKudoTableVisitor<T, R> implements SchemaVisitor<T, R> {
+    private final List<KudoTable> tables;
+    private final long[] currentValidityOffsets;
+    private final long[] currentOffsetOffsets;
+    private final long[] currentDataOffset;
+    private final Deque<SliceInfo>[] sliceInfoStack;
+    private final Deque<Long> totalRowCountStack;
```

Comment:
We cannot support a row count larger than an int.  I'm fine while we're accumulating to use a long to capture overflow conditions (e.g.: within the constructor when it accumulates and checks for overflow).  I see no reason to track computed  total rows that are larger than an int because the overflow check should have already happened in the constructor and in  updateOffsets which is missing an overflow check.